### PR TITLE
BUG: fix float16 type not being called due to wrong ordering

### LINKED ISSUE #7897: np.sqrt.types is bogus
Notice how some loops are duplicated, and `e->e` comes after `f->f`, ensuring it never gets selected.

```
>>> np.__version__
'1.11.1'
>>> np.sqrt
<ufunc 'sqrt'>
>>> np.sqrt.types

['f->f',
 'd->d',
 'e->e',
 'f->f',
 'd->d',
 'g->g',
 'F->F',
 'D->D',
 'G->G',
 'O->O']
```

### PR DESCRIPTION
**Pull request by juliantaylor**
closes gh-7897

### CODE REVIEW

**Comment by juliantaylor** [comment]
doesn't fix the ugliness in scalarmath, maybe someone has good idea how to do this

**Comment by juliantaylor** [comment]
interestingly pow does test this, I ran into they in my exp,log,sin,cos,pow vectorization branch too

File: numpy/core/src/umath/scalarmath.c.src [hunk 1]
```diff
@@ -1703,9 +1703,10 @@ get_functions(PyObject * mm)
      * generate_umath.py, the first to go into FLOAT/DOUBLE_sqrt
      * they have the same signature as the scalar variants so we need to skip
      * over them
+     * also skip float16 copy placed before
      */
-    i = 4;
-    j = 2;
+    i = 6;
```
Comment (charris) [inline numpy/core/src/umath/scalarmath.c.src]: Aha... Boy, that *is* ugly ;)

File: numpy/linalg/tests/test_linalg.py [hunk 2]
```diff
@@ -877,11 +877,11 @@ def test_vector_return_type(self):
 
             an = norm(at, 2)
             assert_(issubclass(an.dtype.type, np.floating))
-            assert_almost_equal(an, 2.0**(1.0/2.0))
+            assert_almost_equal(an, an.dtype.type(2.0)**an.dtype.type(1.0/2.0))
```
Comment (charris) [inline numpy/linalg/tests/test_linalg.py]: Out of curiousity, is this because of accuracy, or some other reason.

**Comment by charris** [comment]
Thanks Julian.

**Comment by charris** [comment]
I've been thinking of getting rid of the ufuncs in `scalarmath` and just using the `npymath` lib. I already got rid of some of the ufuncs for `divmod`, just didn't finish the job.

**Pull request merged**